tsdb/shard.go

```diff
@@ -178,26 +286,143 @@ func (s *Shard) WritePoints(points []Point) error {
+			id, _ := wal.NextSequence()
+
+			// Append points sequentially to the WAL bucket.
+			v := marshalWALEntry(p.Key(), p.UnixNano(), p.Data())
+			if err := wal.Put(u64tob(id), v); err != nil {
+				return fmt.Errorf("put wal: %s", err)
+			}
+		}
+
+		return nil
+	}); err != nil {
+		_ = s.Close()
+		return err
+	}
+
+	// If successful then save points to in-memory cache.
+	if err := func() error {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		for _, p := range points {
+			// Generate in-memory cache entry of <timestamp,data>.
+			key, data := p.Key(), p.Data()
+			v := make([]byte, 8+len(data))
+			binary.BigEndian.PutUint64(v[0:8], uint64(p.UnixNano()))
+			copy(v[8:], data)
+
+			// TODO(benbjohnson): Check that we're not inserting a duplicate.
+
+			// Determine if we are appending.
+			a := s.cache[string(key)]
+			appending := (len(a) == 0 || bytes.Compare(a[len(a)-1], v) == -1)
+
+			// Add entry to cache.
+			a = append(a, v)
+
+			// Sort by timestamp if not appending.
+			if !appending {
+				sort.Sort(byteSlices(a))
+			}
+
+			s.cache[string(key)] = a
+
+			// Calculate estimated WAL size.
+			s.walSize += len(key) + len(v)
+		}
+
+		// Check for flush threshold.
+		s.autoflush()
+
+		return nil
+	}(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Flush writes all points from the write ahead log to the index.
+func (s *Shard) Flush() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if err := s.db.Update(func(tx *bolt.Tx) error {
+		// Iterate over keys in the WAL bucket.
+		c := tx.Bucket([]byte("wal")).Cursor()
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			key, timestamp, data := unmarshalWALEntry(v)
+
+			// Create bucket for entry.
+			b, err := tx.CreateBucketIfNotExists(key)
 			if err != nil {
-				return err
+				return fmt.Errorf("create bucket: %s", err)
 			}
-			if err := bp.Put(u64tob(uint64(p.UnixNano())), p.Data()); err != nil {
-				return err
+
+			// Write point to bucket.
+			if err := b.Put(u64tob(uint64(timestamp)), data); err != nil {
+				return fmt.Errorf("put: %s", err)
+			}
+
+			// Remove entry in the WAL.
+			if err := c.Delete(); err != nil {
+				return fmt.Errorf("delete: %s", err)
 			}
 		}
 
+		// Reset WAL size.
+		s.walSize = 0
+
 		return nil
 	}); err != nil {
-		_ = s.Close()
 		return err
 	}
 
+	// Reset cache.
+	s.cache = make(map[string][][]byte)
+
 	return nil
 }
 
+// autoflusher waits for notification of a flush and kicks it off in the background.
+// This method runs in a separate goroutine.
+func (s *Shard) autoflusher(closing chan struct{}) {
+	defer s.wg.Done()
+
+	for {
+		// Wait for close or flush signal.
+		select {
+		case <-closing:
+			return
+		case <-s.flush:
+			if err := s.Flush(); err != nil {
+				s.logger.Printf("flush error: %s", err)
+			}
+		}
+	}
+}
+
+// autoflush signals that a flush should occur if the size is above the threshold.
+// This function must be called within the context of a lock.
+func (s *Shard) autoflush() {
```

Comment:
Small thing, but I don't like the name of this function. The use of "auto" is misleading, as it implies something automatically happens. I was expecting to see something else in here. A name like "flushThreshold", "flushSize", something, would be better IMHO. Not sure, but something about his name is off to me.